File: models/cates.js
var types = require('types');
var db = require('./models/db.js');
var dbUtils = require('./models/dbUtils.js');

var db = db.getConn();



exports.getTree = function getTree($rows, $pid, $level) {
  $pid = $pid || 0;
  $level = $level || 0;

  var $tree = [];
  for (var k in $rows) {
    var $row = $rows[k];
    if ($row.parent_id == $pid) {
      $tree.push({  //$row是模型实体对象无法添加level属性
        oid: $row.oid,
        cat_name: $row.cat_name,
        intro: $row.intro,
        parent_id: $row.parent_id,
        level: $level
      });
      //if($row.oid != $row.parent_id){ //原本以为会死循环，但实际上不会出现
      $tree = $tree.concat(getTree($rows, $row.oid, $level + 1));
      //}
    }
  }
  return $tree;
}

exports.getFamily = function getFamily($rows, $catid) {
  var $arr = [], k, row, isFind;
  while ($catid != 0) {
    isFind = false;
    for (k in $rows) {
      row = $rows[k];
      if (row.oid == $catid) {
        $arr.unshift(row);
        $catid = row.parent_id;
        isFind = true;//避免死循环
        break;
      }
    }
    if (!isFind) break;
  }
  return $arr;
}

exports.getChildCates = function getChilds($rows, $catid) {
  var arr = [], k, r;
  for (k in $rows) {
    r = $rows[k];
    if (r.parent_id == $catid) {
      arr.push(r.oid);
      arr = arr.concat(getChilds($rows, r.oid));
    }
  }
  return arr;
}

exports.find = function find() {
  var ret = [];
  var r = db.query('select * from cates');
  if (r.err) {
    throw r.err;
  }
  var rows = r.value;
  while (rows.next()) {
    var id = types.newInt();
    var oid = types.newString();
    var cat_name = types.newString();
    var intro = types.newString();
    var parent_id = types.newString();
    var created_at = types.newString();

    var err = rows.scan(id, oid, cat_name, intro, parent_id, created_at);
    if (err) {
      rows.close();
      throw err;
    }

    ret.push({
      id: types.intValue(id),
      oid: types.stringValue(oid),
      cat_name: types.stringValue(cat_name),
      intro: types.stringValue(intro),
      parent_id: types.stringValue(parent_id),
      created_at: types.stringValue(created_at),
    });
  }
  var err = rows.err();
  if (err) {
    rows.close();
    throw err;
  }
  err = rows.close();
  if (err) {
    throw err;
  }
  return ret;
}

exports.add = function add(data) {
  var r = dbUtils.exportKeyValues(data);
  var sql = dbUtils.makeInsertSql('cates', r.keys);
  return dbUtils.insert(db, sql, r.values);
}

exports.delByOid = function delByOid(oid) {
  return dbUtils.delete(db, 'delete from cates where oid=?', [oid]);
}

exports.getByOid = function getByOid(oid) {
  var ret = null;
  var r = db.query('select * from cates where oid = ?', oid);
  if (r.err) {
    throw r.err;
  }
  var rows = r.value;
  while (rows.next()) {
    var id = types.newInt();
    var oid = types.newString();
    var cat_name = types.newString();
    var intro = types.newString();
    var parent_id = types.newString();
    var created_at = types.newString();

    var err = rows.scan(id, oid, cat_name, intro, parent_id, created_at);
    if (err) {
      throw err;
    }

    ret = {
      id: types.intValue(id),
      oid: types.stringValue(oid),
      cat_name: types.stringValue(cat_name),
      intro: types.stringValue(intro),
      parent_id: types.stringValue(parent_id),
      created_at: types.stringValue(created_at),
    };
    break;
  }
  var err = rows.err();
  if (err) {
    throw err;
  }
  err = rows.close();
  if (err) {
    throw err;
  }
  return ret;
}

exports.updateByOid = function updateByOid(oid, data) {
  var r = dbUtils.exportKeyValues(data)
  var sql = dbUtils.makeUpdateSql('cates', r.keys, ['oid'])
  // 添加where条件值
  r.values.push(oid)
  return dbUtils.update(db, sql, r.values)
}

File: models/cates.go
package models

import (
	"database/sql"
)

type CatesModel struct {
	db *sql.DB
}

func NewCatesModel(db *sql.DB) *CatesModel {
	return &CatesModel{db}
}

type CateTreeItem struct {
	Oid       string
	Cat_name  string
	Intro     string
	Parent_id string
	Level     int
}

func (c *CatesModel) GetTree(rows []*Cate) []*CateTreeItem {
	return c.getTree(rows, "0", 0)
}

func (c *CatesModel) getTree(rows []*Cate, pid string, level int) []*CateTreeItem {
	var tree = make([]*CateTreeItem, 0)
	for _, row := range rows {
		if row.Parent_id == pid {
			tree = append(tree, &CateTreeItem{
				Oid:       row.Oid,
				Cat_name:  row.Cat_name,
				Intro:     row.Intro,
				Parent_id: row.Parent_id,
				Level:     level,
			})
			tree = append(tree, c.getTree(rows, row.Oid, level+1)...)
		}
	}
	return tree
}

func (c *CatesModel) GetFamily(rows []*Cate, catid string) []*Cate {
	var arr = make([]*Cate, 0)
	var isFind bool
	for catid != "0" {
		isFind = false
		for _, row := range rows {
			if row.Oid == catid {
				var tmp = make([]*Cate, len(arr)+1)
				tmp[0] = row
				copy(tmp[1:], arr)
				arr = tmp
				catid = row.Parent_id
				isFind = true //避免死循环
				break
			}
		}
		if !isFind {
			break
		}
	}
	return arr
}

func (c *CatesModel) GetChildCates(rows []*Cate, catid string) []string {
	var arr = make([]string, 0)
	for _, r := range rows {
		if r.Parent_id == catid {
			arr = append(arr, r.Oid)
			arr = append(arr, c.GetChildCates(rows, r.Oid)...)
		}
	}
	return arr
}

func (c *CatesModel) Find() []*Cate {
	r, err := c.Query("select * from cates")
	if err != nil {
		panic(err)
	}
	return r.([]*Cate)
}

type Cate struct {
	Id         int
	Oid        string
	Cat_name   string
	Intro      string
	Parent_id  string
	Created_at string
}

func (c *CatesModel) Query(sql string, params ...interface{}) (interface{}, error) {
	rows, err := c.db.Query(sql, params...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var ret = make([]*Cate, 0)
	for rows.Next() {
		cate := &Cate{}
		var err = rows.Scan(&cate.Id, &cate.Oid, &cate.Cat_name, &cate.Intro, &cate.Parent_id, &cate.Created_at)
		if err != nil {
			return nil, err
		}
		ret = append(ret, cate)
	}
	err = rows.Err()
	if err != nil {
		return nil, err
	}
	return ret, nil
}

func (c *CatesModel) Add(data map[string]interface{}) *DMLResult {
	var r = ExportKeyValues(data)
	var sql = MakeInsertSql("cates", r.Keys)
	var ret, err = DML(c.db, sql, r.Values...)
	if err != nil {
		panic(err)
	}
	return ret
}

func (c *CatesModel) DelByOid(oid string) *DMLResult {
	r, err := DML(c.db, "delete from cates where oid=?", oid)
	if err != nil {
		panic(err)
	}
	return r
}

func (c *CatesModel) GetByOid(oid string) []*Cate {
	r, err := c.Query("select * from cates where oid = ?", oid)
	if err != nil {
		panic(err)
	}
	return r.([]*Cate)
}

func (c *CatesModel) UpdateByOid(oid string, data map[string]interface{}) *DMLResult {
	var r = ExportKeyValues(data)
	var sql = MakeUpdateSql("cates", r.Keys, "oid")
	var ret, err = DML(c.db, sql, append(r.Values, oid)...)
	if err != nil {
		panic(err)
	}
	return ret
}
